File: python/OneImputedFamilyRef.py
from __future__ import annotations

# coding: utf-8
# OneImputedFamilyRef.py
# 片親がimputeされていてもう片親はunknownな家系を補完する

from typing import Optional

from VCF import VCFSmall
from VCFGeno import VCFGenoBase, VCFGeno
from VCFFamily import *
import OneImputedFamily
import RefCommon
from KnownFamily import *
from OptionSmall import OptionSmall


#################### OneImputedFamilyRef ####################

def impute(orig_vcf: VCFSmall, phased_vcf: VCFGeno, ref_haps: list[list[int]],
									families: list[KnownFamily],
									op: OptionSmall) -> Optional[VCFGenoBase]:
	vcfs: list[VCFGenoBase] = []
	set_imputed_samples = set(phased_vcf.samples)
	for family in families:
		is_mat_imputed = family.mat in set_imputed_samples
		vcf = VCFGeno.extract_samples(family.samples(), orig_vcf);
		records = RefCommon.merge_family_records(phased_vcf,
													vcf, family.samples())
		vcf1 = OneImputedFamily.create_family_vcf(family, is_mat_imputed,
													records, len(families),
													ref_haps, orig_vcf, op)
		vcf1.impute()
		vcfs.append(vcf1)
	
	if not vcfs:
		return None
	
	print("%d families whose one parent is imputed and the other parent is"
								" unknown have been imputed." % len(families))
	new_vcf = VCFGeno.join(vcfs, orig_vcf.samples)
	return new_vcf

__all__ = ['impute']

File: src/OneImputedFamilyRef.cpp
#include "../include/OneImputedFamilyRef.h"
#include "../include/OneImputedFamily.h"
#include "../include/VCFImputable.h"
#include "../include/KnownFamily.h"
#include "../include/OptionSmall.h"
#include "../include/common.h"
#include "../include/RefCommon.h"

using namespace std;


//////////////////// OneImputedeFamilyRef ////////////////////

VCFGeno *OneImputedFamilyRef::impute(const VCFSmall *orig_vcf,
									const VCFGeno *imputed_vcf,
									const vector<vector<int>>& ref_haps,
									const vector<const KnownFamily *>& families,
									const OptionSmall& op) {
	const size_t	L = families.size();
	vector<VCFImputable *>	vcfs;
	vector<VCFGeno *>	vcf_garbage;
	const vector<string>&	imputed_samples = imputed_vcf->get_samples();
	const set<string>	set_imputed_samples(imputed_samples.begin(),
											imputed_samples.end());
	for(auto p = families.begin(); p != families.end(); ++p) {
		const KnownFamily	*family = *p;
		const STRVEC&	samples = family->get_samples();
		const bool	is_mat_imputed =
							set_imputed_samples.find(family->get_mat())
											!= set_imputed_samples.end();
		auto	*vcf = VCFGeno::extract_samples(samples, orig_vcf);
		auto	records = RefCommon::merge_family_records(imputed_vcf,
															vcf, samples);
		auto	*vcf1 = OneImputedFamily::create_family_vcf(family,
													is_mat_imputed,
													records, L,
													ref_haps, orig_vcf, op);
		vcfs.push_back(vcf1);
		vcf_garbage.push_back(vcf);
	}
	
	if(vcfs.empty())
		return NULL;
	
	// Small VCFs are heavy to process, so it will be parallelized.
	VCFImputable::impute_VCFs(vcfs, op.num_threads);
	cout << vcfs.size() << " families whose one parent is imputed and"
				<< " the other parent is unknown have been imputed." << endl;
	vector<const VCFGenoBase *>	vcfs1(vcfs.begin(), vcfs.end());
	auto	*new_vcf = VCFGeno::join(vcfs1, orig_vcf->get_samples());
	Common::delete_all(vcf_garbage);
	Common::delete_all(vcfs);
	return new_vcf;
}
